### ISSUE TITLE
platform-server errors with ivy-ngcc - UMD module - Angular 9.0.0-rc.7

### ISSUE THREAD

**Issue by naveedahmed1** [angular/angular#34544]
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/platform-server

### Is this a regression?

Yes, the previous version in which this bug was not present was: @angular/platform-server@9.0.0-rc.6

### Description

When trying to compile my application with Ivy, I get errors on ivy ngcc for platform-server when compiling for umd bundles (for SSR)

## 🔬 Minimal Reproduction
Update any project to latest version of Angular (9.0.0-rc.7).

Add following config to `angular.json`:

```
"server": {
          "builder": "@angular-devkit/build-angular:server",
          "options": {
            "outputPath": "dist-server",
            "main": "src/main.server.ts",
            "tsConfig": "src/tsconfig.server.json"
          },
          "configurations": {
            "production": {
              "optimization": true,
              "outputHashing": "media",
              "sourceMap": false,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "bundleDependencies": "all",
              "fileReplacements": [
                {
                  "replace": "src/environments/environment.ts",
                  "with": "src/environments/environment.prod.ts"
                }
              ]
            }
          }
        }
```

Then run command:

`ng run MyProject:server:production`

## 🔥 Exception or Error
<pre><code>
Compiling @angular/platform-server : main as umd
Hash: f0b9625353a4d2f5af81
Time: 12116ms
Built at: 12/23/2019 8:02:36 PM
  Asset      Size  Chunks  Chunk Names
main.js  1.75 KiB       0  main
Entrypoint main = main.js
chunk    {0} main.js (main) 28 bytes [entry] [rendered]

ERROR in Expected UMD module file (/node_modules/url/node_modules/punycode/punycode.js) to contain exactly one statement, but found [object Object],[object Object].
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.7
Node: 12.13.1
OS: win32 x64

Angular: 9.0.0-rc.7
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router, service-worker
Ivy Workspace: Yes

Package                                    Version
--------------------------------------------------------------------
@angular-devkit/architect                  0.900.0-rc.7
@angular-devkit/build-angular              0.900.0-rc.7
@angular-devkit/build-optimizer            0.900.0-rc.7
@angular-devkit/build-webpack              0.900.0-rc.7
@angular-devkit/core                       9.0.0-rc.7
@angular-devkit/schematics                 9.0.0-rc.7
@angular/cdk                               9.0.0-rc.5
@angular/fire                              5.2.1
@angular/google-maps                       9.0.0-rc.5
@angular/material                          9.0.0-rc.5
@angular/pwa                               0.900.0-rc.7
@ngtools/webpack                           9.0.0-rc.7
@nguniversal/module-map-ngfactory-loader   9.0.0-next.9
@schematics/angular                        9.0.0-rc.7
@schematics/update                         0.900.0-rc.7
rxjs                                       6.5.3
typescript                                 3.6.4
webpack                                    4.41.2
</code></pre>

**Comment by naveedahmed1** [comment]
I am not sure but I believe it is similar to https://github.com/angular/angular/issues/31069

**Comment by naveedahmed1** [comment]
@petebacondarwin @kara I am not sure but I believe it is similar to https://github.com/angular/angular/issues/31069

**Comment by gkalpak** [comment]
@naveedahmed1, I cannot reproduce the issue. Here is what I tried:

```ts
# Attempt 1
npm install --global @angular/cli@9.0.0-rc.7
ng new test-app --defaults
cd test-app
ng add @nguniversal/express-engine --client-project test-app
ng run test-app:server:production

# Attempt 2
rm -rf node_modules
npm install
# Copied your config from https://github.com/angular/angular/issues/34544#issue-541795169 into `angular.json`
ng run test-app:server:production

# Attempt 3
./node_modules/.bin/ngcc
```

In all cases above, the commands completed successfully and in the last one `@angular/platform-server` was compiled as UMD (saw `Compiling @angular/platform-server : main as umd` in the logs).

##
Can you, please, provide a minimal reproduction (e.g. a project we cach check out and run) with specific instructions on how to reproduce the problem?

**Comment by naveedahmed1** [comment]
Can I add you to a private repo?

**Comment by naveedahmed1** [comment]
@gkalpak I have added you to a private repo and dropped you an email.

**Comment by gkalpak** [comment]
Thx, @naveedahmed1. I took a look and there are actually two issues involved:

1. `@angular/platform-server` imports Node.js' built-in [`url` module][1] [here][2]. `ngcc` fails to recognize that this is the Node.js built-in and tries to process [`url` package][3] (which is a transitive dependency of the project and ends up in top-level `node_modules/` via npm deduping.

2. The [`url package][3] dependes on the [`punycode` package][4], whose [main file][4] is in the form `;(function(root) { ... }(this));` (which is neither CommonJS nor UMD). `ngcc` assumes that the file pointed to by a `package.json`'s `main` property will be in either CommonJS or UMD format. Given that `punycode.js` does not look like CommonJS, `ngcc` tries to parse it as UMD and it expects UMD files to be in format `(function (factory) { ... })(function (...) { ... });` (and thus only have one top-level statement). However, the file has two top-level statements (one empty statement, followed by a parenthesized function call): `;(function(root) { ... }(this));`

##
Having 100% static evaluation for the UMD format is probably out of scope for `ngcc` (as it would be quite complex/costly with all the different variations), so we have to rely on heuristics (which can't be 100% accurate). Off the top of my head, here are some things we can improve:

1. Teach `ngcc` to ignore Node.js built-in modules as dependencies.
2. Do not assume that `package.json > main` points to either CommonJS or UMD. Improve our detection of the format `main` files and throw a more descriptive error for unsupported formats.

[1]: https://nodejs.org/api/url.html
[2]: https://github.com/angular/angular/blob/9b9116c79/packages/platform-server/src/location.ts#L12
[3]: https://www.npmjs.com/package/url
[4]: https://www.npmjs.com/package/punycode
[5]: https://unpkg.com/browse/punycode@1.4.1/punycode.js

**Comment by naveedahmed1** [comment]
Thank you so much @gkalpak for sharing the details. Let me also share that it works fine with 'Angular 9.0.0-rc.6' so if you could please also take a look at what exactly has changed in Angular 9.0.0-rc.7 that causes ngcc to throw this error.

**Comment by petebacondarwin** [comment]
BTW, we already ignore built-in modules: https://github.com/angular/angular/blob/master/packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts#L149

**Comment by tayambamwanza** [comment]
Can confirm also experiencing this issue only from Angular 9.0.0-rc.7, did not exist in Angular 9.0.0-rc.6

**Comment by Gbuomprisco** [comment]
I'm experiencing a similar problem, caused by a dependency of https://github.com/ngneat/transloco

```
Compiling @ngneat/transloco-utils : main as umd
Error: Error on worker #3: Error: Expected UMD module file (/Users/giancarlo/Lab/project/node_modules/@ngneat/transloco-utils/node_modules/cosmiconfig/dist/loaders.js) to contain exactly one statement, but found [object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object].
    at UmdReflectionHost.getUmdModule (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/host/umd_host.js:61:27)
    at UmdReflectionHost.getModuleStatements (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/host/umd_host.js:113:34)
    at UmdReflectionHost.Esm2015ReflectionHost.findClassSymbols (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/host/esm2015_host.js:427:18)
    at DecorationAnalyzer.analyzeFile (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/analysis/decoration_analyzer.js:139:55)
    at /Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/analysis/decoration_analyzer.js:129:59
    at Array.map (<anonymous>)
    at DecorationAnalyzer.analyzeProgram (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/analysis/decoration_analyzer.js:129:18)
    at Transformer.analyzeProgram (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/packages/transformer.js:123:57)
    at Transformer.transform (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/packages/transformer.js:69:27)
    at ClusterWorker.compile (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/main.js:162:42)
    at ClusterMaster.onWorkerMessage (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/execution/cluster/master.js:158:27)
    at /Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/execution/cluster/master.js:46:95
    at ClusterMaster.<anonymous> (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/execution/cluster/master.js:238:57)
    at step (/Users/giancarlo/Lab/project/node_modules/tslib/tslib.js:136:27)
    at Object.next (/Users/giancarlo/Lab/project/node_modules/tslib/tslib.js:117:57)
    at /Users/giancarlo/Lab/project/node_modules/tslib/tslib.js:110:75
    at new Promise (<anonymous>)
    at Object.__awaiter (/Users/giancarlo/Lab/project/node_modules/tslib/tslib.js:106:16)
    at EventEmitter.<anonymous> (/Users/giancarlo/Lab/project/node_modules/@angular/compiler-cli/ngcc/src/execution/cluster/master.js:232:32)
    at EventEmitter.emit (events.js:200:13)
Compiling @angular/compiler/testing : es2015 as esm2015
Compiling @angular/core : module as esm5
Compiling @angular/core : main as umd
Compiling @angular/core : es2015 as esm2015
```

**Comment by naveedahmed1** [comment]
It seems that this issue is caused by the recent changes related to UMD bundles in Angular 9.0.0-rc.7

https://github.com/angular/angular/pull/34356

https://github.com/angular/angular/commit/84a7d8a

https://github.com/angular/angular/commit/c26738d

**Comment by gkalpak** [comment]
This seems like a different problem, @Gbuomprisco. Can you please open a separate issue (if the problem is still present in 9.0.0-rc.8)?

**Comment by Gbuomprisco** [comment]
Hi @gkalpak, the issues above are resolved. 

I'm still unable to compile due to https://github.com/angular/angularfire/issues/2280 but I guess this has to be solved in Angular Fire.

**Comment by gkalpak** [comment]
Cool! Thx for the update.

**Comment by naveedahmed1** [comment]
I'm still having this issue with latest `Angular 9.0.0-rc.8`.

**Comment by tayambamwanza** [comment]
> Hi @gkalpak, the issues above are resolved.
> 
> I'm still unable to compile due to [angular/angularfire#2280](https://github.com/angular/angularfire/issues/2280) but I guess this has to be solved in Angular Fire.

Could you link the new angularfire issue so that we can follow it getting resolved there?

**Comment by kainiedziela** [comment]
Issue not solved with `AngulaR 9.0.0-rc.9`.

**Comment by gkalpak** [comment]
@naveedahmed1, @kainiedziela (or anyone else having the same issue): Could you try with the [build artifacts][1] from #34811 and see if the problem is fixed for you?

[1]: https://circleci.com/gh/angular/angular/589268#artifacts/containers/0

**Comment by naveedahmed1** [comment]
@gkalpak sure I can try that can you please guide how to use it?

BTW I also added you to a private repository, if you could please use that to test it? Use `npm run build:ssr` to reproduce the error.

**Comment by gkalpak** [comment]
> sure I can try that can you please guide how to use it?

You can find instructions on how to use them [here][1]. Basically, you just replace the versions in your package.json to point to the artifacts URL.

> BTW I also added you to a private repository, if you could please use that to test it? Use `npm run build:ssr` to reproduce the error.

I have done that already and the error goes away. But I want to verify that everything still works as expected at runtime and I am not sure how to do that for your app.

[1]: https://github.com/angular/angular/blob/f2df1c79d/docs/DEVELOPER.md#getting-packages-from-build-artifacts

**Comment by naveedahmed1** [comment]
@gkalpak I tried it and its throwing below error:

An unhandled exception occurred: compiler_1.Version is not a constructor
See "C:\Users\NAVEED~1\AppData\Local\Temp\ng-rz0OsH\angular-errors.log" for further details.

I am using below packages:

```
  "dependencies": {
    "@angular/animations": "https://589268-24195339-gh.circle-artifacts.com/0/angular/animations-pr34811-38e9d9ef1a.tgz",
    "@angular/cdk": "9.0.0-rc.5",
    "@angular/common": "https://589268-24195339-gh.circle-artifacts.com/0/angular/common-pr34811-38e9d9ef1a.tgz",
    "@angular/compiler": "https://589268-24195339-gh.circle-artifacts.com/0/angular/compiler-cli-pr34811-38e9d9ef1a.tgz",
    "@angular/core": "https://589268-24195339-gh.circle-artifacts.com/0/angular/core-pr34811-38e9d9ef1a.tgz",
    "@angular/fire": "5.2.1",
    "@angular/forms": "https://589268-24195339-gh.circle-artifacts.com/0/angular/forms-pr34811-38e9d9ef1a.tgz",
    "@angular/google-maps": "9.0.0-rc.5",
    "@angular/material": "9.0.0-rc.5",
    "@angular/platform-browser": "https://589268-24195339-gh.circle-artifacts.com/0/angular/platform-browser-pr34811-38e9d9ef1a.tgz",
    "@angular/platform-browser-dynamic": "https://589268-24195339-gh.circle-artifacts.com/0/angular/platform-browser-dynamic-pr34811-38e9d9ef1a.tgz",
    "@angular/platform-server": "https://589268-24195339-gh.circle-artifacts.com/0/angular/platform-server-pr34811-38e9d9ef1a.tgz",
    "@angular/pwa": "0.900.0-rc.6",
    "@angular/router": "https://589268-24195339-gh.circle-artifacts.com/0/angular/router-pr34811-38e9d9ef1a.tgz",
    "@angular/service-worker": "https://589268-24195339-gh.circle-artifacts.com/0/angular/service-worker-pr34811-38e9d9ef1a.tgz",
    "@nguniversal/module-map-ngfactory-loader": "9.0.0-next.9",

```

```
  "devDependencies": {
    "@angular-builders/custom-webpack": "8.4.1",
    "@angular-devkit/build-angular": "^0.900.0-rc.9",
    "@angular/cli": "^9.0.0-rc.7",
    "@angular/compiler-cli": "https://589268-24195339-gh.circle-artifacts.com/0/angular/compiler-cli-pr34811-38e9d9ef1a.tgz",
    "@angular/language-service": "https://589268-24195339-gh.circle-artifacts.com/0/angular/language-service-pr34811-38e9d9ef1a.tgz",

```

**Comment by petebacondarwin** [comment]
This is your problem @naveedahmed1 :

```
"@angular/compiler": "https://589268-24195339-gh.circle-artifacts.com/0/angular/compiler-cli-pr34811-38e9d9ef1a.tgz",
```

Notice that the URL refers to `compiler-cli`.

**Comment by naveedahmed1** [comment]
Thank you so much @petebacondarwin :)

> I have done that already and the error goes away. But I want to verify that everything still works as expected at runtime and I am not sure how to do that for your app.

@gkalpak yes the backend is in .Net and that's a separate repo. I tested it on my side and I can confirm that the issue has been resolved on runtime as well. Thank you so much for the quick fix :)

Looking forward to have it in next release.

**Comment by marcnoon** [comment]
Was able to compile with very rudimentary workaround in

**Comment by marcnoon** [comment]
Was able to compile with yarn install using basic work around: 

![image](https://user-images.githubusercontent.com/883172/72690874-33d0fa00-3ad5-11ea-82df-b63bf6fc6e35.png)

**Comment by naveedahmed1** [comment]
> I'm curious to know - is @angular/fire with bundleDependencies all working?

@Gbuomprisco it was working fine with Angular 9.0.0-rc.6 and earlier. Hopefully @gkalpak's  recent commit (https://github.com/gkalpak/angular/commit/8e53a2b034969633147f75b73c8352b066668f50) will fix this in next release.

**Comment by fergalmoran** [comment]
Does anyone know if this commit made it into RC10? It's a bit painful having to use the build artifacts.

**Comment by gkalpak** [comment]
No, the fix hasn't landed yet. This is the PR (which is still open): #34811

**Comment by naveedahmed1** [comment]
@AndrewKushnir can you please confirm if its fixed in RC 11?

**Comment by gkalpak** [comment]
Yes, the fix is included in v9.0.0-rc.11.

**Comment by naveedahmed1** [comment]
Thank you so much @gkalpak for the confirmation :)
